typing/typemod.ml

```diff
@@ -707,6 +708,77 @@ let merge_constraint initial_env loc sg lid constr =
   with Includemod.Error explanation ->
     raise(Error(loc, initial_env, With_mismatch(lid.txt, explanation)))
 
+(* A simplified version of [merge_constraint] for the case of packages. Package
+   constraints are much simpler - they must be "with type" constraints for types
+   with no parameters and can only update abstract types. *)
+let merge_package_constraint initial_env loc sg lid cty =
+  let rec patch_item namelist outer_sig_env sg_for_env ~ghosts item =
+    let return replace_by =
+      Some ((), {Signature_group.ghosts; replace_by})
+    in
+    match item, namelist with
+    | Sig_type(id, sig_decl, rs, priv) , [s]
+      when Ident.name id = s ->
+        begin match sig_decl.type_manifest with
+        | None -> ()
+        | Some ty ->
+          raise (Error(loc, outer_sig_env, With_package_manifest (lid.txt, ty)))
+        end;
+        let new_sig_decl =
+          Typedecl.transl_package_constraint ~loc outer_sig_env cty.ctyp_type
+        in
+        check_type_decl outer_sig_env sg_for_env loc id None
+          new_sig_decl sig_decl;
+        let new_sig_decl = { new_sig_decl with type_manifest = None } in
+        return (Some(Sig_type(id, new_sig_decl, rs, priv)))
+    | Sig_module(id, _, md, rs, priv) as item, s :: namelist
+      when Ident.name id = s ->
+        let sig_env = Env.add_signature sg_for_env outer_sig_env in
```

Comment:
I'm not convinced this is necessary -- seems like it shouldn't be.

Reply:
Updating the environment?  Here's an example that needs it:

```ocaml
module type S = sig
  module type S1 = sig
    type t
  end
  module M : S1
end

type t = (module S with type M.t = int)
```

If we don't extend the environment with the definiton of `S1`, expanding `M`'s type fails on the next line.  I've added this to the test suite.

(Admittedly this took me a bit to work out.)

Reply:
Ah, yes, good example!